### ISSUE TITLE
Securing WSL - How to have basic security for WSL?

### ISSUE THREAD

**Issue by ghost** [Microsoft/WSL#1538]
I typed in Google "How to secure WSL" (without quote marks) and found no "general-audience-friendly" article that gives some tips about basic security setups for WSL.

I personally want to use WSL to manage my SSH connections, and consider using WSL-OpenSSH instead of Puttygen-Pageant-Putty.

I wondered how should I secure a basic WSL installment (basically just WSL and OpenSSH and some totally-local utilites like Tree or Zip --- No Webserver, no databases).

I encourage people to create a short, accessible article on this in a blog... Would gladly read it.

**Comment by fpqc** [comment]
Right now it is insecure by design due to interop. All users are "really" root users.

**Comment by aseering** [comment]
@fpqc -- well, what's your threat model?  All users may be root users, but root is not a Windows Administrator.

**Comment by fpqc** [comment]
@aseering True, but my threat model is a Linux instance acting as a persistent server as Benosika seems to be indicating.

**Comment by Benosika** [comment]
Would you agree that OpenSSh server isn't included in the servers you meant to?

**Comment by dmex** [comment]
@fpqc 

"It's firewalled by Windows, so should be safe anyhow I think."

Pico processes are unaffected by firewall rules and are able to send and receive data from the network even when blocked by firewall software (due to the lack of ImageName for pico processes #1031).

**Comment by ionescu007** [comment]
@dmex That's actually not correct. Pico processes *are* affected by firewall rules, and additionally in Creators Update, the name query API returns the name for WSL processes.

**Comment by pnegry** [comment]
@ionescu007 reference for this? how do you create windows firewall rules for pico processes?

**Comment by Benosika** [comment]
What about securing an OpenSSH server? I need one on WSL to transfer files easy and fast from my VPS --- To the local session (WSL). Any tips?

**Comment by aseering** [comment]
@Benosika -- can you use the ssh server built into your VPS?

**Comment by Benosika** [comment]
Yes.

**Comment by Benosika** [comment]
I do wonder what a WSL user that has an OpenSSH-server and uses only a password (without key pairs) could do to reduce the chance for Burte force attack, if it can't install firewalls (something which as of the moment, generally isn't possible in WSL). @aseering

**Comment by aseering** [comment]
As discussed on another ticket, I wonder if DenyHosts works?

**Comment by Benosika** [comment]
`sudo apt-get install denyhosts -y` doesn't work in WSL, it outputs "Unable to locate package". Yet, in a native Ubuntu it does being installed fine.

**Comment by dmex** [comment]
@Benosika 
> What about securing an OpenSSH server?

Windows automatically installs an SSH server when you enable the Developer mode option in the Win10 Settings app. Have you tried connecting to the machine and running bash.exe?

> could do to reduce the chance for Burte force attack

The SSH server uses local Windows accounts and generates logon/logoff audits and other events in the Event log. The main problem with the default SSH server on Win10 (security wise) seems to be that each connection to the ssh server creates a new process and it doesn't limit the number of connections.

**Comment by Benosika** [comment]
@dmex , right after I've enabled the developer tools I installed WSL and there I had an existing OpenSSH server. Does it say that I now have 2 OpenSSH servers on my Windows10? One of Windows itself and one of WSL?

**Comment by aseering** [comment]
@Benosika -- denyhosts is in Ubuntu's "universe" repository; you'll have to make sure that it is enabled in `/etc/apt/sources.list`.

**Comment by aseering** [comment]
fail2ban might work too, though.  I'd say give it a try.  That configuration looks right to me, though I haven't configured fail2ban recently so I'm not an expert.

Also, don't forget to start it manually -- just like ssh itself, denyhosts or fail2ban won't be started automatically by Windows.

**Comment by benhillis** [comment]
Thanks for the discussion.

**Comment by xOrMalware** [comment]
This is an interesting thread. The question that springs to mind, is what is the security between the subsystem and host windows system? Is WSL being considered as VM, and thus, their should not be any leaks of data between host and subsystem - and their is no concept of breakout or "break in" for that matter.

Maybe I am thinking too deeply and comparing the system to a VM, but in my eyes, there needs to be sometype of subsystem security to prevent leaks of data. Hypothetical example, you have RDP enabled on the host windows. You are hacked, by a questionable sort. If the person is not able to log in to the Linux subsystem are there protections/mitigations in place to prevent memory reads, disk reads etc.

The only reason I ask, is that I am going to look at a breakout scenario for a VM - and my mind wondered on to what WSL has to offer in this area.
